src/Tools/dotnet-monitor/Exceptions/ConfiguredExceptionsStore.cs

```diff
@@ -0,0 +1,138 @@
+        {
+        }
+
+        internal ConfiguredExceptionsStore(int topLevelLimit, ExceptionsStoreCallback callback = null)
+        {
+            if (topLevelLimit <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(topLevelLimit));
+            }
+
+            _store = new ExceptionsStore(new Callback(this, topLevelLimit, callback));
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            return _store.DisposeAsync();
+        }
+
+        public void AddExceptionInstance(IExceptionsNameCache cache, ulong exceptionId, ulong groupId, string message, DateTime timestamp, ulong[] stackFrameIds, int threadId, ulong[] innerExceptionIds, string activityId, ActivityIdFormat activityIdFormat)
+        {
+            _store.AddExceptionInstance(cache, exceptionId, groupId, message, timestamp, stackFrameIds, threadId, innerExceptionIds, activityId, activityIdFormat);
+        }
+
+        public void RemoveExceptionInstance(ulong exceptionId)
+        {
+            _store.RemoveExceptionInstance(exceptionId);
+        }
+
+        public IReadOnlyList<IExceptionInstance> GetSnapshot()
+        {
+            return _store.GetSnapshot();
+        }
+
+        private sealed class Callback : ExceptionsStoreCallback
+        {
+            private readonly ExceptionsStoreCallback _callback;
+            private readonly Dictionary<ulong, List<ulong>> _innerExceptionMap = new();
+            private readonly Dictionary<ulong, List<ulong>> _outerExceptionMap = new();
+            private readonly ConfiguredExceptionsStore _store;
+            private readonly LinkedList<ulong> _topLevelExceptions = new();
+            private readonly int _topLevelLimit;
+
+            public Callback(ConfiguredExceptionsStore store, int topLevelLimit, ExceptionsStoreCallback callback)
+            {
+                _callback = callback;
+                _store = store;
+                _topLevelLimit = topLevelLimit;
+            }
+
+            public override void AfterAdd(IExceptionInstance instance)
+            {
+                _callback?.AfterAdd(instance);
+            }
+
+            public override void BeforeAdd(IExceptionInstance instance)
+            {
+                _outerExceptionMap.Add(instance.Id, new List<ulong>());
+                _innerExceptionMap.Add(instance.Id, instance.InnerExceptionIds.ToList());
+
+                // Make sure all of the inner exceptions are no longer top-level exceptions.
+                foreach (ulong innerExceptionId in instance.InnerExceptionIds)
+                {
+                    if (_outerExceptionMap.TryGetValue(innerExceptionId, out List<ulong> outerExceptions))
+                    {
+                        outerExceptions.Add(instance.Id);
+                    }
+
+                    _topLevelExceptions.Remove(innerExceptionId);
+                }
+
+                _topLevelExceptions.AddLast(instance.Id);
```

Comment:
I'll preface this by saying this is purely hypothetical (and I'm not sure if it matters) - I'm wondering if it's more performant to add to the front. My understanding is this is a DLL with head/tail pointers, _but_ the `Remove` algorithm removes the first occurrence of the value (i.e. iterating from the head pointer). This could be an incorrect assumption, but I'd guess that outer exceptions often become inner exceptions shortly after we process them (i.e., if we had a limit of 1000, we'd add exception_1 to the back...and then shortly after we'd add it's outer exception_2, meaning we now need to traverse all the way to the back to remove something that was just added). If we add to the front, searching with `Remove` in the average case (with these assumptions) would probably be faster. You can also remove at the back or front for when we hit the limit, so it shouldn't be an issue to switch to `RemoveLast`.

As previously mentioned - might be making some incorrect assumptions, and didn't run any benchmarks, but thought I'd throw it out there since that was my first thought reading through this.

Reply:
Absolutely right! This was probably a remnant of when I was using list-based tracking, but forgot to update the algorithm.